src/mmc_windows.c

```diff
@@ -33,163 +33,162 @@ void mmc_finalize()
+ */
+void dismount_volume(LPWSTR volume_name) {
+    HANDLE volume_handle = CreateFile(volume_name,
                                       GENERIC_READ | GENERIC_WRITE,
                                       FILE_SHARE_READ | FILE_SHARE_WRITE,
                                       NULL,
                                       OPEN_EXISTING,
                                       0,
                                       NULL);
     if (volume_handle == INVALID_HANDLE_VALUE) {
-        fwup_warnx("Cannot open '%s'", mmc_device);
-        return -1;
+        fwup_errx(EXIT_FAILURE, "Error dismounting '%S', (Error %lu)", volume_name, GetLastError());
+        return;
     }
 
     DWORD bytes_returned;
     BOOL status = DeviceIoControl(volume_handle,
-                             FSCTL_LOCK_VOLUME,
-                             NULL,
-                             0,
-                             NULL,
-                             0,
-                             &bytes_returned,
-                             NULL);
+                                  FSCTL_DISMOUNT_VOLUME,
+                                  NULL,
+                                  0,
+                                  NULL,
+                                  0,
+                                  &bytes_returned,
+                                  NULL);
+    CloseHandle(volume_handle);
+
     if (!status) {
-        fwup_warnx("Error locking '%s'", mmc_device);
-        CloseHandle(volume_handle);
-        return -1;
+        fwup_errx(EXIT_FAILURE, "Error dismounting '%S', (Error %lu)", volume_name, GetLastError());
     }
+}
 
-    status = DeviceIoControl(volume_handle,
-                             FSCTL_DISMOUNT_VOLUME,
-                             NULL,
-                             0,
-                             NULL,
-                             0,
-                             &bytes_returned,
-                             NULL);
-    if (!status) {
-        DWORD error = GetLastError();
-        if (error != ERROR_NOT_SUPPORTED) {
-            fwup_warnx("Error locking '%s'", mmc_device);
-            CloseHandle(volume_handle);
-            return -1;
-        } else {
-            fwup_warnx("Unmounting not supported");
-        }
+/*
+ * @brief Unmount all LogicalVolumes using the specified PhysicalDrive
+ * @param mmc_device the name of the PhysicalDrive
+ * @return 0 on success, exit the program with a failure message otherwise
+ */
+int mmc_umount_all(const char *mmc_device)
+{
+    unsigned int target_disk_number = 0;
+    sscanf(mmc_device, "\\\\.\\PhysicalDrive%u", &target_disk_number);
+    if (target_disk_number == 0)
+        fwup_errx(EXIT_FAILURE, "Target device must be formatted like \\\\.\\PhysicalDisk# where # is a positive integer.");
+
+    WCHAR volume_name[MAX_PATH] = L"";
+    HANDLE volume_iter = FindFirstVolume(volume_name, ARRAYSIZE(volume_name));
+
+    if (volume_iter == INVALID_HANDLE_VALUE) {
```

Comment:
No brace for one line blocks. I mostly follow the Qt coding style in my C and C++ projects: http://wiki.qt.io/Qt_Coding_Style.